pytorch/1. The Torch Tensor Library and Basic Operations.ipynb

```diff
@@ -0,0 +1,585 @@
+   "source": [
+    "print (x.size())\n",
+    "y = x + torch.randn(5, 1)\n",
+    "print(y)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "collapsed": true
+   },
+   "source": [
+    "### Reshape"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 14,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "torch.Size([5, 10, 15])\n",
+      "torch.Size([50, 15])\n",
+      "torch.Size([50, 1, 15])\n",
+      "torch.Size([50, 15])\n",
+      "\n",
+      "torch.Size([10, 5, 15])\n",
+      "torch.Size([5, 15, 10])\n",
+      "torch.Size([10, 15, 5])\n",
+      "torch.Size([10, 15, 5])\n"
+     ]
+    }
+   ],
+   "source": [
+    "y = torch.randn(5, 10, 15)\n",
+    "print(y.size())\n",
+    "print(y.view(-1, 15).size())  # Same as doing y.view(50, 15)\n",
+    "print(y.view(-1, 15).unsqueeze(1).size()) # Adds a dimension at index 1.\n",
+    "print(y.view(-1, 15).unsqueeze(1).squeeze().size())\n",
+    "# If input is of shape: (Ax1xBxCx1xD)(Ax1xBxCx1xD) then the out Tensor will be of shape: (AxBxCxD)(AxBxCxD)\n",
+    "print()\n",
+    "print(y.transpose(0, 1).size())\n",
+    "print(y.transpose(1, 2).size())\n",
+    "print(y.transpose(0, 1).transpose(1, 2).size())\n",
+    "print(y.permute(1, 2, 0).size())"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Repeat"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 15,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "torch.Size([50, 100, 15])\n",
+      "torch.Size([50, 100, 15])\n"
+     ]
+    }
+   ],
+   "source": [
+    "print(y.view(-1, 15).unsqueeze(1).expand(50, 100, 15).size())\n",
+    "print(y.view(-1, 15).unsqueeze(1).expand_as(torch.randn(50, 100, 15)).size())"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Concatenate"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 16,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "torch.Size([5, 10, 30])\n",
+      "torch.Size([2, 5, 10, 15])\n"
+     ]
+    }
+   ],
+   "source": [
+    "# 2 is the dimension over which the tensors are concatenated\n",
+    "print(torch.cat([y, y], 2).size())\n",
+    "# stack concatenates the sequence of tensors along a new dimension.\n",
+    "print(torch.stack([y, y], 0).size())"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Advanced Indexing"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 17,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "torch.Size([2, 3])\n",
+      "torch.Size([2, 3, 4])\n"
+     ]
+    }
+   ],
+   "source": [
+    "y = torch.randn(2, 3, 4)\n",
+    "print(y[:, 1, :3].size())\n",
```

Comment:
That's not advanced indexing, though.
The simplest kind of advanced indexing is indexing by a tensor, for instance `print(y[[1, 0, 1, 1]])` (shape would be (4, 3, 4)).
More complex can be something like `print(y[[0, 1], [[1, 0], [2, 1]], [[3], [1]]])`, where the indices are broadcasted together and iterated jointly on (the shape here would be (2, 2), corresponding to indices (0, 1, 3), (1, 0, 3), (0, 2, 1), (1, 1, 1)).